core/src/server/rpc_module.rs

```diff
@@ -80,6 +83,45 @@ pub type MaxResponseSize = usize;
 ///   - a [`mpsc::UnboundedReceiver<String>`] to receive future subscription results
 pub type RawRpcResponse = (String, mpsc::Receiver<String>);
 
+#[derive(Debug, Clone)]
+#[allow(missing_copy_implementations)]
+/// The connection details exposed to the server methods.
+pub struct ConnectionDetails {
+	id: ConnectionId,
+}
+
+impl ConnectionDetails {
+	/// Construct a new [`ConnectionDetailsBuilder`].
+	pub fn builder() -> ConnectionDetailsBuilder {
+		ConnectionDetailsBuilder { id: 0 }
+	}
+
+	/// Get the connection ID.
+	pub fn id(&self) -> ConnectionId {
+		self.id
+	}
+}
+
+#[derive(Debug, Clone)]
+#[allow(missing_copy_implementations)]
+/// The connection details exposed to the server methods.
```

Comment:
The `RpcService` should be the last "layer" before the user-callback is called.

We construct the `ConnectionDetails` in this layer:

https://github.com/paritytech/jsonrpsee/blob/87585ad948dcdbd7a31773ba8305f38cc9c21fbe/server/src/middleware/rpc/layer/rpc_service.rs#L102-L104

I think we could just as easily provide a constructor for this: 

```rust
impl ConnectionDetails {
  pub fn new(id: ConnectionId) ..
}
```

The `new` method needs to be public, because the [ConnectionDetails](https://github.com/paritytech/jsonrpsee/blob/87585ad948dcdbd7a31773ba8305f38cc9c21fbe/core/src/server/rpc_module.rs#L89-L91) is declared in the `core` crate. And the `ConnectionDetails` needs to be instantiated in the `server` crate as well.

We'd have to eventually change this method to `new(id: ConnectionId, extensions: Extenstions)` which will be a breaking change (although we should be the only ones building this object).

TLDR; since we'll do a breaking change either way, I think a `new` method should be fine for now! 🙏

Reply:
ok, I see makes sense

Reply:
Maybe have it be `#[doc(hidden)]` and comment that it's not part of the public APi and can change without notice?